### DISCUSSION TITLE
GFW Resist TLS Proxy

### DISCUSSION THREAD

**Discussion by MortThePersian** [XTLS/Xray-core#1988] [Q&A]
Dear RPRX;

I just came across this article,

https://github.com/GFW-knocker/gfw_resist_tls_proxy

& I felt it had to be shared with you ASAP !

Sincerely

**Comment by us254** [comment]
https://github.com/ValdikSS/GoodbyeDPI

https://github.com/SadeghHayeri/GreenTunnel

https://github.com/nemolian/nosni_proxy

**Reply by MortThePersian** [reply]
May The God's richest blessings shines upon the Knights who combat the darkness

**Comment by RPRX** [comment]
这种方式完全适用我十天前的一条评论：https://github.com/XTLS/Xray-core/issues/1906#issuecomment-1512176531

简单来说，可以用，但不适合大范围推广，若引起了 GFW 的注意，很快它就会被检测到

因为正常的 TLS 流量并不会把 Client Hello 分成多个帧发送，找不到任何合理的解释，GFW 只需检测这一点即可抓到它

**Reply by GibMeMyPacket** [reply]
That `gfw_resist_tls_proxy` idea made a short-term hope for Iranians but there are several to many reports that it doesn't work in some specific ISPs, even thought it's shared earlier.

Reality is the one and only hope for Iranians, it's being used almost widely and no one yet reported that it doesn't work on a specific ISP with the correct configuration / good IP.

But recently it's (actually I'm) facing `net/http: TLS handshake timeout` time to time in an specific ISP, especially at the peak of internet usage! This is like about 2 URL tests ends up with that error at those times (more at mornings) and the whole for proxy doesn't work till i get about 3 `net/http: TLS handshake timeout`
Tried several to many SNI, all the same.

This is more likely that GFW is finding a pattern or such between requests, but this is weird that it will reconnect and work for another 5/10 URL tests.

are there any updates going to happen on making REALITY more and more REAL?

**Reply by RPRX** [reply]
> But recently it's (actually I'm) facing `net/http: TLS handshake timeout` time to time in an specific ISP, especially at the peak of internet usage! This is like about 2 URL tests ends up with that error at those times (more at mornings) and the whole for proxy doesn't work till i get about 3 `net/http: TLS handshake timeout` Tried several to many SNI, all the same.

请提供更多的信息，比如是 Vision 还是 H2 / gRPC，并在下次发生这种情况时看一下服务端有没有收到请求

REALITY 还有一些隐藏玩法，比如目标网站有默认证书时，你可以使用任意 SNI、无 SNI，可以试一试

无 SNI：需要 Xray-core v1.8.1+，服务端 `serverNames` 加一个 `""`，客户端 `address` 为 IP、`serverName` 为 `""` 即可

**Reply by ghost** [reply]
> This approach works perfectly for a comment I made ten days ago: [#1906 (comment)](https://github.com/XTLS/Xray-core/issues/1906#issuecomment-1512176531)
> 
> In short, it can be used, but it is not suitable for large-scale promotion. If it attracts the attention of GFW, it will be detected soon
> 
> Because normal TLS traffic does not send Client Hello into multiple frames, there is no reasonable explanation, and GFW only needs to detect this to catch it

what you think about [Geneva](https://github.com/Kkevsterrr/geneva)?

**Reply by yoohooyoo** [reply]
> > But recently it's (actually I'm) facing `net/http: TLS handshake timeout` time to time in an specific ISP, especially at the peak of internet usage! This is like about 2 URL tests ends up with that error at those times (more at mornings) and the whole for proxy doesn't work till i get about 3 `net/http: TLS handshake timeout` Tried several to many SNI, all the same.
> 
> 请提供更多的信息，比如是 Vision 还是 H2 / gRPC，并在下次发生这种情况时看一下服务端有没有收到请求
> 
> REALITY 还有一些隐藏玩法，比如目标网站有默认证书时，你可以使用任意 SNI、无 SNI，可以试一试
> 
> 无 SNI：需要 Xray-core v1.8.1+，服务端 `serverNames` 加一个 `""`，客户端 `address` 为 IP、`serverName` 为 `""` 即可

话说...任意SNI...是不是可以免流了额

**Reply by deximy** [reply]
> > > But recently it's (actually I'm) facing `net/http: TLS handshake timeout` time to time in an specific ISP, especially at the peak of internet usage! This is like about 2 URL tests ends up with that error at those times (more at mornings) and the whole for proxy doesn't work till i get about 3 `net/http: TLS handshake timeout` Tried several to many SNI, all the same.
> > 
> > 
> > 请提供更多的信息，比如是 Vision 还是 H2 / gRPC，并在下次发生这种情况时看一下服务端有没有收到请求
> > REALITY 还有一些隐藏玩法，比如目标网站有默认证书时，你可以使用任意 SNI、无 SNI，可以试一试
> > 无 SNI：需要 Xray-core v1.8.1+，服务端 `serverNames` 加一个 `""`，客户端 `address` 为 IP、`serverName` 为 `""` 即可
> 
> 话说...任意SNI...是不是可以免流了额

为什么要任意SNI才能免流？偷证书就已经可以免了吧？

**Reply by GibMeMyPacket** [reply]
>
> 请提供更多的信息，比如是 Vision 还是 H2 / gRPC，并在下次发生这种情况时看一下服务端有没有收到请求
> 

[VLESS-XTLS-uTLS-REALITY](https://github.com/XTLS/REALITY#readme)
I will update this topic once i was sure that those disconnections are because of xray and not my internet itself!

> REALITY 还有一些隐藏玩法，比如目标网站有默认证书时，你可以使用任意 SNI、无 SNI，可以试一试
> 
> 无 SNI：需要 Xray-core v1.8.1+，服务端 `serverNames` 加一个 `""`，客户端 `address` 为 IP、`serverName` 为 `""` 即可

However, emtpy SNI doesn't work (tried both v1.8.1 and [latest master commit](https://github.com/XTLS/Xray-core/actions/runs/4778065779))!
`[Warning] [NNN] app/proxyman/outbound: failed to process outbound traffic > proxy/vless/outbound: failed to find an available destination > common/retry: [x509: cannot validate certificate for *.*.*.* because it doesn't contain any IP SANs] > common/retry: all retry attempts failed`
server config is as follow:
```
              "dest": "www.yahoo.com:443",
              "xver": 0,
              "serverNames": [
                  "yahoo.com",
                  "www.yahoo.com",
                  ""
              ],
```

client config:
```
      "streamSettings": {
        "network": "tcp",
        "security": "reality",
        "realitySettings": {
          "serverName": "",
          "spiderX": "/"
...
```

**Reply by RPRX** [reply]
@GibMeMyPacket 

1. 请检查 yahoo.com 是否有“默认证书”：https://t.me/projectXtls/78
2. 若有，报错大概率是因为你没有重启服务端

**Reply by laaArmy** [reply]
> > > But recently it's (actually I'm) facing `net/http: TLS handshake timeout` time to time in an specific ISP, especially at the peak of internet usage! This is like about 2 URL tests ends up with that error at those times (more at mornings) and the whole for proxy doesn't work till i get about 3 `net/http: TLS handshake timeout` Tried several to many SNI, all the same.
> > 
> > 请提供更多的信息，比如是 Vision 还是 H2 / gRPC，并在下次发生这种情况时看一下服务端有没有收到请求
> > 
> > REALITY 还有一些隐藏玩法，比如目标网站有默认证书时，你可以使用任意 SNI、无 SNI，可以试一试
> > 
> > 无 SNI：需要 Xray-core v1.8.1+，服务端 `serverNames` 加一个 `""`，客户端 `address` 为 IP、`serverName` 为 `""` 即可
> 
> 话说...任意SNI...是不是可以免流了额 

同问，免流复活了，目测

**Reply by A5DkjGQUZx** [reply]
> This approach works perfectly for a comment I made ten days ago: [#1906 (comment)](https://github.com/XTLS/Xray-core/issues/1906#issuecomment-1512176531)
> 
> In short, it can be used, but it is not suitable for large-scale promotion. If it attracts the attention of GFW, it will be detected soon
> 
> Because normal TLS traffic does not send Client Hello into multiple frames, there is no reasonable explanation, and GFW only needs to detect this to catch it

Agree, but can be a tool available for us in our arsenal

**Reply by RPRX** [reply]
https://github.com/XTLS/Xray-core/issues/2000#issuecomment-1528913218

**Reply by seanozman** [reply]
> > > > But recently it's (actually I'm) facing `net/http: TLS handshake timeout` time to time in an specific ISP, especially at the peak of internet usage! This is like about 2 URL tests ends up with that error at those times (more at mornings) and the whole for proxy doesn't work till i get about 3 `net/http: TLS handshake timeout` Tried several to many SNI, all the same.
> > > 
> > > 
> > > 请提供更多的信息，比如是 Vision 还是 H2 / gRPC，并在下次发生这种情况时看一下服务端有没有收到请求
> > > REALITY 还有一些隐藏玩法，比如目标网站有默认证书时，你可以使用任意 SNI、无 SNI，可以试一试
> > > 无 SNI：需要 Xray-core v1.8.1+，服务端 `serverNames` 加一个 `""`，客户端 `address` 为 IP、`serverName` 为 `""` 即可
> > 
> > 
> > 话说...任意SNI...是不是可以免流了额
> 
> 同问，免流复活了，目测

澳洲免流，有人弄过吗？
